go/x509util/x509util.go

```diff
@@ -0,0 +1,576 @@
+	OidExtensionCRLDistributionPoints      = asn1.ObjectIdentifier{2, 5, 29, 31}
+	OidExtensionIssuerAltName              = asn1.ObjectIdentifier{2, 5, 29, 18}
+	OidExtensionSubjectDirectoryAttributes = asn1.ObjectIdentifier{2, 5, 29, 9}
+	OidExtensionInhibitAnyPolicy           = asn1.ObjectIdentifier{2, 5, 29, 54}
+	OidExtensionPolicyConstraints          = asn1.ObjectIdentifier{2, 5, 29, 36}
+	OidExtensionPolicyMappings             = asn1.ObjectIdentifier{2, 5, 29, 33}
+	OidExtensionFreshestCRL                = asn1.ObjectIdentifier{2, 5, 29, 46}
+
+	OidExtensionAuthorityInfoAccess = asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 1}
+	OidExtensionSubjectInfoAccess   = asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 11}
+
+	// RFC 6962 s3.1
+	OidExtensionCTPoison = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 3}
+	// RFC 6962 s3.3
+	OidExtensionCTSCT = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 2}
+)
+
+//////////////////////////////////////
+
+// OidForStandardExtension indicates whether oid identifies a standard extension.
+// Standard extensions are listed in RFC 5280 (and other RFCs).
+func OidForStandardExtension(oid asn1.ObjectIdentifier) bool {
+	if oid.Equal(OidExtensionSubjectKeyId) ||
+		oid.Equal(OidExtensionKeyUsage) ||
+		oid.Equal(OidExtensionExtendedKeyUsage) ||
+		oid.Equal(OidExtensionAuthorityKeyId) ||
+		oid.Equal(OidExtensionBasicConstraints) ||
+		oid.Equal(OidExtensionSubjectAltName) ||
+		oid.Equal(OidExtensionCertificatePolicies) ||
+		oid.Equal(OidExtensionNameConstraints) ||
+		oid.Equal(OidExtensionCRLDistributionPoints) ||
+		oid.Equal(OidExtensionIssuerAltName) ||
+		oid.Equal(OidExtensionSubjectDirectoryAttributes) ||
+		oid.Equal(OidExtensionInhibitAnyPolicy) ||
+		oid.Equal(OidExtensionPolicyConstraints) ||
+		oid.Equal(OidExtensionPolicyMappings) ||
+		oid.Equal(OidExtensionFreshestCRL) ||
+		oid.Equal(OidExtensionSubjectInfoAccess) ||
+		oid.Equal(OidExtensionAuthorityInfoAccess) ||
+		oid.Equal(OidExtensionCTPoison) ||
+		oid.Equal(OidExtensionCTSCT) {
+		return true
+	}
+	return false
+}
+
+// OidInExtensions indicates whether the extension identified by oid is present in extensions.
+func OidInExtensions(oid asn1.ObjectIdentifier, extensions []pkix.Extension) (int, bool) {
+	count := 0
+	critical := false
+	for _, ext := range extensions {
+		if ext.Id.Equal(oid) {
+			count++
+			if ext.Critical {
+				critical = true
+			}
+		}
+	}
+	return count, critical
+}
+
+// String formatting for various X.509/ASN.1 types
+func bitStringToString(b asn1.BitString) string {
+	result := hex.EncodeToString(b.Bytes)
+	bitsLeft := b.BitLength % 8
+	if bitsLeft != 0 {
+		result += " (" + strconv.Itoa(8-bitsLeft) + " unused bits)"
+	}
+	return result
+}
+
+func publicKeyAlgorithmToString(algo x509.PublicKeyAlgorithm) string {
+	switch algo {
+	case x509.RSA:
+		return "rsaEncryption"
+	case x509.DSA:
+		return "dsaEncryption"
+	case x509.ECDSA:
+		return "id-ecPublicKey"
```

Comment:
Is there something special about ECDSA that means that the returned string for it should be of a different sort of format to the other two?  I don't know loads about crypto algos, or the purpose you have in mind for this function (although I'm sure it's further down in this file), so this may be completely fine, was just curious :)